backend/test_observer/controllers/reports/test_executions.py

```diff
@@ -0,0 +1,120 @@
+import csv
+from datetime import datetime
+from typing import Any
+
+from fastapi import APIRouter, Depends
+from fastapi.responses import FileResponse
+from sqlalchemy import select, func, text, Select
+from sqlalchemy.orm import Session
+
+from test_observer.data_access.models import (
+    Artefact,
+    ArtefactBuild,
+    ArtefactBuildEnvironmentReview,
+    Environment,
+    Family,
+    Stage,
+    TestEvent,
+    TestExecution,
+)
+from test_observer.data_access.setup import get_db
+
+
+router = APIRouter()
+
+TEST_EXECUTIONS_REPORT_COLUMNS: list[Any] = [
+    Family.name,
+    Artefact.id,
+    Artefact.name,
+    Artefact.version,
+    Artefact.status,
+    Artefact.track,
+    Artefact.series,
+    Artefact.repo,
+    TestExecution.id,
+    TestExecution.status,
+    TestExecution.ci_link,
+    TestExecution.c3_link,
+    TestExecution.checkbox_version,
+    TestExecution.created_at,
+    Environment.name,
+    Environment.architecture,
+    ArtefactBuildEnvironmentReview.review_decision,
+    ArtefactBuildEnvironmentReview.review_comment,
+    text("test_executions_events.testevents"),
+]
+
+
+def _get_test_executions_reports_query(
+    start_date: datetime, end_date: datetime
+) -> Select:
+    """
+    Builds the query that retrieves the test executions based on the parameters set
+    """
+    test_events_subq = (
+        select(
+            TestExecution.id.label("test_execution_id"),
+            func.coalesce(
+                func.array_agg(
+                    func.json_build_object(
+                        "event_name",
+                        TestEvent.event_name,
+                        "timestamp",
+                        func.to_char(TestEvent.timestamp, "YYYY-MM-DD HH24:MI:SS:MS"),
+                        "detail",
+                        TestEvent.detail,
+                    ),
+                ).filter(TestEvent.id.is_not(None)),
+                [],
+            ).label("testevents"),
+        )
+        .outerjoin(TestEvent)
+        .group_by(TestExecution.id)
+        .alias("test_executions_events")
+    )
+
+    return (
+        select(*TEST_EXECUTIONS_REPORT_COLUMNS)
+        .join_from(TestExecution, Environment)
+        .join_from(TestExecution, ArtefactBuild)
+        .join_from(ArtefactBuild, Artefact)
+        .join_from(Artefact, Stage)
+        .join_from(Stage, Family)
+        .join(
+            ArtefactBuildEnvironmentReview,
+            (ArtefactBuildEnvironmentReview.artefact_build_id == ArtefactBuild.id)
+            & (ArtefactBuildEnvironmentReview.environment_id == Environment.id),
+        )
+        .join(
+            test_events_subq,
+            TestExecution.id == test_events_subq.c.test_execution_id,
+        )
+        .where(
+            TestExecution.created_at >= start_date, TestExecution.created_at <= end_date
+        )
+    )
```

Comment:
I've proposed another query [here](https://github.com/canonical/test_observer/pull/231). See if it makes sense

Reply:
Turns out the alternative query had a bug. So nvm that